### PR TITLE
Merge multi-sel movements patches

### PR DESCRIPTION
**Pull request by raphlinus**
#243 was accidentally merged into the wrong branch. I think this is the way to merge it into master.

Progress towards #188

### CODE REVIEW

File: rust/core-lib/src/movement.rs [hunk 1]
```diff
@@ -25,43 +29,158 @@ pub enum Movement {
...
 
-/// Calculate a horizontal position in the view, based on the offset. Return
-/// value has the same type as `region_movement` for convenience.
-fn calc_horiz(view: &View, text: &Rope, offset: usize) -> (usize, Option<HorizPos>) {
-    let (_line, col) = view.offset_to_line_col(text, offset);
-    (offset, Some(col))
+/// Computes the actual desired amount of scrolling (generally slightly
+/// less than the height of the viewport, to allow overlap).
+fn scroll_height(view: &View) -> isize {
+    max(view.scroll_height() as isize - 2, 1)
```
Comment (cmyr) [inline rust/core-lib/src/movement.rs]: is `2` the 'slop' constant?

File: rust/core-lib/src/movement.rs [hunk 2]
```diff
@@ -25,43 +29,158 @@ pub enum Movement {
...
+            let mut cursor = Cursor::new(&text, r.end);
+            let offset = cursor.prev::<LinesMetric>().unwrap_or(0);
+            (offset, None)
+        }
+        Movement::EndOfParagraph => {
+            // Note: TextEdit would start at modify ? r.end : r.max()
+            let mut offset = r.end;
+            let mut cursor = Cursor::new(&text, offset);
+            if let Some(next_para_offset) = cursor.next::<LinesMetric>() {
+                if cursor.is_boundary::<LinesMetric>() {
```
Comment (cmyr) [inline rust/core-lib/src/movement.rs]: to clarify: if `cursor.is_boundary() == false` in this situation we're at the end of the file?

File: rust/core-lib/src/movement.rs (hunk 1 above)
Comment (raphlinus) [inline rust/core-lib/src/movement.rs]: No, this is something else; it's the overlap (number of lines in common) on successive page scrolls. As slop is increased, this should stay the same. It should become a constant with a comment though.

File: rust/core-lib/src/movement.rs (hunk 2 above)
Comment (raphlinus) [inline rust/core-lib/src/movement.rs]: Yes.

**Comment by raphlinus** [comment]
I want `flags` to go away, everything should be captured in enum parameters to the movement command.

**Pull request merged**